flash/data/data_source.py

```diff
@@ -0,0 +1,263 @@
+        data_pipeline: DataPipeline,
+    ) -> Optional[Union[AutoDataset, IterableAutoDataset]]:
+        mock_dataset = MockDataset()
+        with CurrentRunningStageFuncContext(running_stage, "load_data", self):
+            data = self.load_data(data, mock_dataset)  # TODO: Should actually resolve this
+
+        if has_len(data):
+            dataset = AutoDataset(data, self, running_stage, data_pipeline)
+        else:
+            dataset = IterableAutoDataset(data, self, running_stage, data_pipeline)
+        dataset.__dict__.update(mock_dataset.metadata)
+        return dataset
+
+
+class DefaultDataSource(Enum):  # TODO: This could be replaced with a data source registry that the user can add to
+
+    FOLDERS = "folders"
+    FILES = "files"
+
+    def as_type(self) -> Type[DataSource]:
+        _data_source_types = {
+            DefaultDataSource.FOLDERS: FoldersDataSource,
+            DefaultDataSource.FILES: FilesDataSource,
+        }
+        return _data_source_types[self]
+
+
+class SequenceDataSource(DataSource, ABC):
+
+    def __init__(
+        self,
+        train_inputs: Optional[Sequence[Any]] = None,
+        train_targets: Optional[Sequence[Any]] = None,
+        val_inputs: Optional[Sequence[Any]] = None,
+        val_targets: Optional[Sequence[Any]] = None,
+        test_inputs: Optional[Sequence[Any]] = None,
+        test_targets: Optional[Sequence[Any]] = None,
+        predict_inputs: Optional[Sequence[Any]] = None,
+        labels: Optional[Sequence[str]] = None
+    ):
+        super().__init__(
+            train_data=(train_inputs, train_targets),
+            val_data=(val_inputs, val_targets),
+            test_data=(test_inputs, test_targets),
+            predict_data=(predict_inputs, None),
+        )
+
+        self.labels = labels
+
+        if self.labels is not None:
+            self.preprocess.set_state(LabelsState(self.labels))
+
+    def load_data(self, data: Any, dataset: Optional[Any] = None) -> Iterable[Mapping[str, Any]]:
+        inputs, targets = data
+        if targets is None:
+            return [{'input': input} for input in inputs]
+        return [{'input': input, 'target': target} for input, target in zip(inputs, targets)]
+
+
+class FoldersDataSource(DataSource, ABC):
+
+    def __init__(
+        self,
+        train_folder: Optional[Union[str, pathlib.Path, list]] = None,
+        val_folder: Optional[Union[str, pathlib.Path, list]] = None,
+        test_folder: Optional[Union[str, pathlib.Path, list]] = None,
+        predict_folder: Optional[Union[str, pathlib.Path, list]] = None,
+        extensions: Optional[Tuple[str, ...]] = None,
+    ):
+        super().__init__(
+            train_data=train_folder,
+            val_data=val_folder,
+            test_data=test_folder,
+            predict_data=predict_folder,
+        )
+
+        self.extensions = extensions
+
+    @staticmethod
+    def find_classes(dir: str) -> Tuple[List[str], Dict[str, int]]:
```

Comment:
shouldn't this be specific to classification tasks ?

Reply:
If it can't get the classes then it will just yield the files and no targets so would all still work fine. It gets a bit complicated if we factor this out into a `ClassificationFoldersDataSource` but we may need to, we'll know more once I add support for the other tasks.

Reply:
add a todo here to move this later